Feature fidsel performance

coreyostrove opened the pull request:
This branch implements a new greedy search algorithm that leverages the same low-rank update tools now available for germ selection and FPR. This is specifically useful for accelerating fiducial selection for many-qubit systems in the 4+ qubit range (for 3 or less the existing algorithms seem to run quick enough for most applications). This new greedy search algorithm only works with an objective function based on the psuedoinverse-trace. Other notable changes in this branch:

- Added additional flexibility to the specification of candidate fiducial sets. Before we could only specify a maximum fiducial length and then all circuits up to that length would be used in the search space. Now the API matches the one used for germ selection were we can specify a dictionary with keys corresponding to lengths and the ability to either provide special keywords like 'all upto' or a value in which case we sample a certain number of circuits at random. We still have compatibility with the old kwarg though.
- There is a new deduping routine that can be used when the gate set members are generators of the clifford group (or some subset thereof).  So gate sets like XYI or XYCPHASE etc... In this case a flag can be set which allows deduping the circuit list in amortized O(N) time.
- Cleaned up a bunch of extraneous debug and print statements throughout various parts of the experiment design codebase (not just in fidsel)
- Fleshed out incomplete germ selection docstring.

sserita commented on pygsti/algorithms/fiducialselection.py:
This might look cleaner if clean_fid_list just took the assume_clifford flag, and switched on it inside the drop_duplicates block.

coreyostrove replied:
Sounds good. I refactored the deduping code implementing this change.

sserita commented on pygsti/algorithms/fiducialselection.py:
```diff
@@ -115,20 +147,48 @@ def find_fiducials(target_model, omit_identity=True, eq_thresh=1e-6,
-                                                                        eq_thresh=eq_thresh)
+        #Use a flag to check if we can assume these are clifford circuits in which case
+        #we can use a more efficient deduping routine.
+        if assume_clifford:
```
Could be cleaned up IMO - see below comment in clean_fid_list_clifford.

coreyostrove replied:
Sounds good, I refactored this in line with your suggestions.

sserita reviewed (changes requested):
Haven't checked the math for the new greedy code path closely, given it's the same process as the recent FPR and germ sel low rank updates.
I think there is one place where code can be consolidated, i.e. we can dedup the dedup code.

enielse commented on pygsti/algorithms/fiducialselection.py:
Maybe make this into a warning that get's issued whenever `max_fid_length` is specified?  This would make the update more visible, just in case people don't read the docstring ;)

coreyostrove replied:
Cool, added an additional warning about this.

enielse commented on pygsti/algorithms/fiducialselection.py:
```diff
@@ -185,6 +244,16 @@ def find_fiducials(target_model, omit_identity=True, eq_thresh=1e-6,
+                if final_test_fiducial_list:
+                    printer.log('Final test of the candidate prep fiducial lists passed.', 1)
+                else:
+                    printer.log('Final test of the candidate prep fiducial lists failed.', 1)
```
The above block is repeated 8 (I think) times within this function, which makes me think it'd be worth consolidating it to a nested function.

coreyostrove replied:
I refactored this as suggested.

enielse commented on pygsti/algorithms/fiducialselection.py:
```diff
@@ -1429,3 +1692,269 @@ def rcl_fn(x): return _scoring.filter_composite_rcl(x, alpha)
+        #if the best fiducial set is not yet initialized
+        #and we have force_empty==True, then we will
+        #initialize it with the index of the empty fiducial.
+        if (best_fiducial_set is None) and force_empty:
```
Maybe, for clarity, change this to `if best_fiducial_set is None:` and place nested `if force_empty: ... else: ...` blocks that over the first two cases here.

coreyostrove replied:
Done.

enielse commented on pygsti/algorithms/germselection.py:
```diff
@@ -147,6 +145,42 @@ def find_germs(target_model, randomize=True, randomization_strength=1e-2,
         If specified this is a number between 0 and 1 that indicates the random fraction of candidate
         germs to drop randomly following the deduping procedure.
+        
+    mode : str, optional (default 'allJac')
```
Should change this doc line to:
`mode : {'allJac', 'singleJac', 'compactEVD'}, optional`
This shows the options clearly while also communicating that it must be a string.  No need to put default on this line, as it's in the function definition and it adds one more place to get things out of sync.

coreyostrove replied:
Done

enielse reviewed (approved):
Great stuff!  I added a few suggestions that you can take or leave.

coreyostrove commented:
Thanks for the comments and feedback, @sserita and @enielse . I have refactored and updated the code in accordance with your suggestions. I'll plan on merging this in once I get a clean set of unit tests back on the latest commit (@sserita, I think you'll also need to approve the PR before merging is unblocked).
